src/example-types/results.js

```diff
@@ -1,79 +1,105 @@
+      throw Error(
+        `The ${targetCollection} schema has a mongo configuration, but doesn't have a 'collection' property`
+      )
 
-    let sort = {
-      [sortField]: sortDir === 'asc' ? 1 : -1,
+    return {
+      $lookup: {
+        as,
+        from: targetCollection,
+        localField: x.localField, // || '_id',
+        foreignField: x.foreignField, // || context.schema, <-- needs schema lookup
+      },
     }
+  })
+
+let getStartRecord = ({ page, pageSize }) => {
+  page = page < 1 ? 0 : page - 1
+  return page * pageSize
+}
 
-    return Promise.all([
-      search(
-        _.reject(_.isEmpty, [
-          {
-            $sort: sort,
-          },
-          {
-            $skip: startRecord,
-          },
-          pageSize > 0 && {
-            $limit: pageSize,
-          },
-          ...F.mapIndexed((x, as) => {
-            let targetSchema = getSchema(x.schema) //|| toSingular(as), //<-- needs compromise-fp
-            if (!targetSchema)
-              throw Error(`Couldn't find schema configuration for ${x.schema}`)
-            if (!targetSchema.mongo)
-              throw Error(
-                'Populating a non mongo provider schema on a mongo provider schema is not supported'
-              )
-            let targetCollection = _.get('mongo.collection', targetSchema)
-            if (!targetCollection)
-              throw Error(
-                `The ${targetCollection} schema has a mongo configuration, but doesn't have a 'collection' property`
-              )
+let getResultsQuery = (context, getSchema, startRecord) => {
+  let { pageSize, sortField, sortDir, populate, include } = context
 
-            return {
-              $lookup: {
-                as,
-                from: targetCollection,
-                localField: x.localField, // || '_id',
-                foreignField: x.foreignField, // || context.schema, <-- needs schema lookup
-              },
-            }
-          }, populate),
-        ])
-      ),
-      search([
-        {
-          $group: {
-            _id: null,
-            count: {
-              $sum: 1,
-            },
-          },
-        },
-      ]),
-    ]).spread((results, count) => ({
-      // TODO - handle aggregate wrapped stuff, e.g. result.result or result.result[0] etc
-      response: {
-        totalRecords: _.get('0.count', count),
-        startRecord: startRecord + 1,
-        endRecord: startRecord + results.length,
-        results,
-      },
-    }))
-  },
+  let $sort = {
+    [sortField]: sortDir === 'asc' ? 1 : -1,
+  }
+
+  // $sort, $skip, $limit
+  let sortSkipLimit = [
+    { $sort },
+    { $skip: startRecord },
+    pageSize > 0 && {
+      $limit: pageSize,
+    },
+  ]
+  // If sort field is a join field move $sort, $skip, and $limit to after $lookup.
+  // Otherwise, place those stages first to take advantage of any indexes on that field.
+  let sortOnJoinField = _.includes(
+    _.replace(/^([^.]+)\..+$/, '$1', sortField),
```

Comment:
This seems a bit complex - could it be something simpler like this?

```js
let sortOnJoinField = _.some(x => _.startsWith(x, sortField), _.keys(populate))
```

Reply:
That doesn't do the same thing.

Reply:
Right, but wouldn't it be all we need? What does this do that wouldn't be handled by startsWith?

Reply:
Here's the problem: https://runkit.com/dubiousdavid/5d14c3f302fba200133e77c6

Reply:
just for reference this was from our slack conversation:

```js
let sortOnJoinField = _.some(
  x => _.startsWith(`${x}.`, sortField) || sortField == x,
  _.keys(populate)
)
```

Reply:
Done.